proposals/0111-linear-types.rst

```diff
@@ -907,6 +909,159 @@ or ``case`` in the surface syntax), we want to infer the multiplicity
+  ``p=e``
+* Either ``p`` is of the form ``!p'`` or ``p`` is a variable. In
+  particular neither ``x@y`` nor ``(x)`` are covered by “is a
+  variable”
+
+If there is a multiplicity annotation, the binding is not
+generalised. So ``let %m x = e in b`` and ``(\(%m x) ->  b) e`` have
+the same typing rule.
+
+When the multiplicity annotation isn't specified, the multiplicity is
+inferred, so that both
+
+::
+
+   \(%1 z) -> let !(x, y) = z in (y, x)
+   let !(x, y) = u in (x, x)
+
+are well typed. Recursive bindings, toplevel bindings and non-variable
+lazy-pattern bindings are always inferred to have multiplicity
+``Many``.
+
+Note that, in particular, function bindings are not allowed:
+
+::
+
+   -- Doesn't work
+   let %p f x y = rhs in body
+
+instead write
+
+::
+
+   -- pat can be a variable
+   let %p f = \x y -> rhs in body
+
+When typechecking ``let %p pat = rhs in body`` the ``rhs`` is consumed
+with multiplicity ``p`` and the variables of ``pat`` must be consumed
+in ``body`` with multiplicity ``p`` (if ``pat`` has some non-linear
+fields, then the variables are scaled appropriately as per
+`Constructors & pattern-matching`_). The pattern ``pat`` can be an
+arbitrary, nested, pattern.
+
+Here are a few examples that illustrate the typing rules
+
+::
+
+   -- good
+   let %1 x = u in x
+
+   -- bad
+   let %1 x = u in (x, x)
+
+   -- bad
+   let %1 x = u in let %Many y = x in …
+
+   -- good
+   let %1 (x, y) = u in (y, x)
+
+   -- bad
+   let %1 (x, y) = u in x
+
+   -- good
+   let %1 (Ur x) = u in (x, x, x)
+
+Non-variable let-bound patterns must be strict
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Non-variable linear patterns can't be lazy (see `Lazy patterns`_). As
+a consequence, non-variable let-bound patterns must be annotated with
+a ``!`` (because let-bound patterns are lazy by default, as opposed to
+case-bound patterns which are strict by default).
+
+::
+
+   -- good
+   let %1 x = u in …
+
+   -- good
+   let %1 !x = u in …
+
+   -- bad
+   let %1 (x, y) = u in …
+
+   -- good
+   let %1 !(x, y) = u in …
+
+   -- inferred unrestricted
+   let (x, y) = u in …
+
+Non-variable linear patterns are monomorphic
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+.. _`Non-variable linear patterns are monomorphic`
+
+Variables in non-variable multiplicity-annotated let-patterns are bound to
+monomorphic types (see `Let bindings and polymorphism`_ for the
+reasoning and a discussion). Unannotated non-variable let-patterns are
+inferred to be unrestricted (by default, since ``-XLinearTypes``
+implies ``-XNoMonoLocalBinds``, only toplevel bindings, which are
+always unrestricted anyway, are inferred to be polymorphic).
+
+::
+
+   {-# LANGUAGE NoMonoLocalBinds #-}
+
+   -- x, y :: Int -> Int
+   let %1 !(x, y) = ((\z->z), (\z->z)) in (x 0, y 1)
+
+   -- x, y :: Int -> Int
+   let %Many !(x, y) = ((\z->z), (\z->z)) in (x 0, x 1, y 2)
+
+   -- x :: forall a. a -> a
+   let %1 x = (\z -> z) in if b then x 1 else x Bool
```

Comment:
Ah! But that's the very point! Linear _non-variable_ pattern bindings aren't generalised. Variable bindings, linear or not, can be freely generalised.

You've missed this point that wasn't meant to be particularly fine. It's worrying. Where do I insist on it to avoid confusion for future reader?